### ISSUE TITLE
panic when using rustdoc on NFS

### ISSUE THREAD

**Issue by ms140569** [rust-lang/rust#40830]
I did:

`RUST_BACKTRACE=1 cargo doc --verbose`

with:

```
rustc 1.17.0-nightly (134c4a0f0 2017-03-20)
cargo 0.19.0-nightly (c995e9eb5 2017-03-17)
```

on:

```
Darwin uranos.fritz.box 14.5.0 Darwin Kernel Version 14.5.0: Sun Sep 25 22:07:15 PDT 2016; root:xnu-2782.50.9~1/RELEASE_X86_64 x86_64
```

and this cargofile:

```toml
[package]
name = "rpx"
version = "0.0.1"
authors = [ "Matthias Schmidt <******>" ]

[dependencies]
getopts = "0.2"
```

and got:

```
 Documenting getopts v0.2.14
     Running `rustdoc --crate-name getopts /Users/schmidtm/.cargo/registry/src/github.com-1ecc6299db9ec823/getopts-0.2.14/src/lib.rs -o /Volumes/server/data/Matthias/prj/rpx/target/doc -L dependency=/Volumes/server/data/Matthias/prj/rpx/target/debug/deps`
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'could not lock `/Volumes/server/data/Matthias/prj/rpx/target/doc/.lock`: Operation not supported (os error 45)', src/librustc_data_structures/flock.rs:362
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: rustc_data_structures::flock::<impl rustc_data_structures::flock::imp::Lock>::panicking_new
   1: rustdoc::html::render::write_shared
   2: rustdoc::html::render::run
   3: rustdoc::rust_input::{{closure}}
   4: std::panicking::try::do_call
   5: __rust_maybe_catch_panic
   6: <F as alloc::boxed::FnBox<A>>::call_box
   7: std::sys::imp::thread::Thread::new::thread_start
   8: _pthread_body
   9: _pthread_start

error: Could not document `getopts`.

Caused by:
  process didn't exit successfully: `rustdoc --crate-name getopts /Users/schmidtm/.cargo/registry/src/github.com-1ecc6299db9ec823/getopts-0.2.14/src/lib.rs -o /Volumes/server/data/Matthias/prj/rpx/target/doc -L dependency=/Volumes/server/data/Matthias/prj/rpx/target/debug/deps` (exit code: 101)
```

HTH,

Matthias Schmidt.

Drop me a mail, if i could help: matthias dot schmidt at gmail dot com

**Comment by sfackler** [comment]
Looks like  your filesystem doesn't support file locks? I know Cargo has a some logic to deal with that which should rustdoc should have as well.

**Comment by nagisa** [comment]
> Looks like your filesystem doesn't support file locks?

> /Volumes/server/data

This looks a lot like a network filesystem (SMB/NFS), which indeed do not support locking.

**Comment by ms140569** [comment]
Yepp, it's a Samba Server on my Linux Box.

Thanks for the evaluation.

Cheers,

Matthias

Am 26.03.2017 10:52 vorm. schrieb "Simonas Kazlauskas" <
notifications@github.com>:

**Comment by areilly** [comment]
FWIW this bug also hits "cargo doc --open" under the same circumstances.  The backtrace shows that this is just running rustdoc, so it's the same thing, just later versions.  I'm up to "guessing_game" in Rust Lang second edition, so it was a bit of a shock to hit breakage so soon.

Yes, /User/andrew/Documents is a symlink to /Volumes/andrew, which is an SMB2 (or 3?) file share on Samba 4.7.4 running on FreeBSD-11.1-STABLE.  I understand that macOS+SMB doesn't support fcntl() locks, and flock() locks are dodgy.  How about using NFS-style dot-locks?

 Documenting libc v0.2.34
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.22.1 (05e2e1c41 2017-11-22) running on x86_64-apple-darwin

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'could not lock `/Volumes/andrew/Documents/rusty_projects/guessing_game/target/doc/.lock`: Operation not supported (os error 45)', src/librustc_data_structures/flock.rs:363:12
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::_print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: std::panicking::begin_panic_fmt
   7: rustc_data_structures::flock::<impl rustc_data_structures::flock::imp::Lock>::panicking_new
   8: rustdoc::html::render::write_shared
   9: rustdoc::html::render::run

error: Could not document `libc`.

Caused by:
  process didn't exit successfully: `rustdoc --crate-name libc /Users/andrew/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.34/src/lib.rs -o /Volumes/andrew/Documents/rusty_projects/guessing_game/target/doc --cfg feature="default" --cfg feature="use_std" -L dependency=/Volumes/andrew/Documents/rusty_projects/guessing_game/target/debug/deps` (exit code: 101)

**Comment by kzys** [comment]
Cargo is ignoring locks on NFS. Maybe we can do the same.

https://github.com/rust-lang/cargo/blob/ce455cd2aa3d11bab3818b8fdee87a11ffceb79b/src/cargo/util/flock.rs#L269

**Comment by kzys** [comment]
Actually I am not so sure why do we need to have this lock.

All crates' docs are generated in parallel, but it would be synchronized at the end of rustdoc execution. Wouldn't be possible to run write_shared() at the end, from the main thread, instead of using flock()?

**Comment by HankB** [comment]
Looks like I ran into the same issue trying to compile with source on an NFS share. (Both host and NFS server running Debian Stretch.)
There are errors in the source but they seem to have nothing to do with the issue (*). `cargo build` ends with 
```text
thread 'main' panicked at 'librustc/session/mod.rs:802: Trying to get session directory from IncrCompSession `NotInitialized`', librustc/session/mod.rs:1296:26
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: aborting due to 3 previous errors


error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.29.1 (b801ae664 2018-09-20) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `tower`.

To learn more, run the command again with --verbose.
hbarta@olive:/mnt/share/projects/tower$ 
```
Impact is minimal for me as I can just move the files to a local filesystem and proceed. If there is any benefit to me providing further information, let me know. (It seems like the source of the problem is already known.)

(*) I've tried code that compiles cleanly and it produces the same error.

**Comment by jettylab** [comment]
Not sure if this is specifically relevant to this issue, but I asked a Rust forum question about randomly failing builds on an NFS mounted volume under MacOS: https://users.rust-lang.org/t/can-rust-builds-be-done-in-an-nfs-mounted-volume/28191.

The advice from @kornel on redirecting `CARGO_TARGET_DIR` is an effective work-around for me for now.

**Comment by jyn514** [comment]
It looks like `flock::new` now returns an error instead of panicking: https://github.com/rust-lang/rust/blob/0c11b93f5a8914a40f619b0a1663baafe029d427/compiler/rustc_data_structures/src/flock.rs#L210. Is anyone still running into this issue?

**Comment by ms140569** [comment]
Hi guys - I'm still facing this issue when compiling anything hosted on my SAMBA/SMB Fileserver:

This is: 
Mac OSX 10.14.6
rustc 1.48.0 (7eac88abb 2020-11-16)

Stacktrace below:

```
   Compiling cynon v0.0.1 (/Volumes/server/data/Matthias/prj/rust/cynon)
     Running `rustc --crate-name build_script_build --edition=2018 build.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C embed-bitcode=no -C debuginfo=2 -C metadata=f8228d0fe8773068 -C extra-filename=-f8228d0fe8773068 --out-dir /Volumes/server/data/Matthias/prj/rust/cynon/target/debug/build/cynon-f8228d0fe8773068 -C incremental=/Volumes/server/data/Matthias/prj/rust/cynon/target/debug/incremental -L dependency=/Volumes/server/data/Matthias/prj/rust/cynon/target/debug/deps --extern prost_build=/Volumes/server/data/Matthias/prj/rust/cynon/target/debug/deps/libprost_build-0de83129d3e4f190.rlib`
error: incremental compilation: could not create session directory lock file: Operation not supported (os error 45)

thread 'rustc' panicked at 'trying to get session directory from `IncrCompSession`: NotInitialized', compiler/rustc_session/src/session.rs:874:48
stack backtrace:
   0: _rust_begin_unwind
   1: std::panicking::begin_panic_fmt
   2: rustc_session::session::Session::incr_comp_session_dir
   3: rustc_incremental::persist::fs::garbage_collect_session_directories
   4: rustc_session::utils::<impl rustc_session::session::Session>::time
   5: rustc_interface::passes::register_plugins
   6: rustc_interface::queries::Queries::register_plugins
   7: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
   8: rustc_span::with_source_map
   9: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.48.0 (7eac88abb 2020-11-16) running on x86_64-apple-darwin

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: aborting due to previous error

error: could not compile `cynon`

Caused by:
  process didn't exit successfully: `rustc --crate-name build_script_build --edition=2018 build.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C embed-bitcode=no -C debuginfo=2 -C metadata=f8228d0fe8773068 -C extra-filename=-f8228d0fe8773068 --out-dir /Volumes/server/data/Matthias/prj/rust/cynon/target/debug/build/cynon-f8228d0fe8773068 -C incremental=/Volumes/server/data/Matthias/prj/rust/cynon/target/debug/incremental -L dependency=/Volumes/server/data/Matthias/prj/rust/cynon/target/debug/deps --extern prost_build=/Volumes/server/data/Matthias/prj/rust/cynon/target/debug/deps/libprost_build-0de83129d3e4f190.rlib` (exit code: 101)
```

**Comment by jyn514** [comment]
@ms140569 that's a separate error - you're running rustc, not rustdoc, and it's an issue with incremental compilation. I'd open a new issue.

I'm going to close this since it seems like the code has been fixed and no one is running into rustdoc errors. Feel free to comment if you're still seeing an issue from rustdoc (and you can't reproduce with `rustc` directly).

**Comment by ms140569** [comment]
Cool, thanks! Sorry for mixing these issues ... Could you send me the issue number against rustc? Cheers, Matthias

**Comment by jyn514** [comment]
@ms140569 https://github.com/rust-lang/rust/issues/76251 looks similar ( I searched `label:A-incr-comp NFS`).
